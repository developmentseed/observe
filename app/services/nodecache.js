import AsyncStorage from '@react-native-community/async-storage'

/**
 *
 * @param {Array<Object>} nodes - array of node objects with lat, lng and id
 * @param {String} tile - tile these nodes fall in
 */
export async function addNodes (tile, nodes) {
  const items = nodes.map(n => [
    `node/${n.id}`,
    JSON.stringify({
      'lat': n.lat,
      'lon': n.lon
    })
  ])
  const nodeIds = nodes.map(n => {
    return `node/${n.id}`
  })

  await AsyncStorage.multiSet(items)
  await AsyncStorage.setItem(tile, JSON.stringify(nodeIds))
}

/**
 * Returns node geometry for an array of node ids
 * @param {Array<Object>} nodeIds - array of node ids
 * @return {<Object>}
 */
export async function getNodes (nodeIds) {
  const nodes = await AsyncStorage.multiGet(nodeIds)
  return nodes.reduce((memo, val) => {
    memo[val[0]] = JSON.parse(val[1])
    return memo
  }, {})
}

/**
 * Get node ids for a tile
 * @param {string} tile - tile id
 * @return {<Object>}
 */
export async function getNodesForTile (tile) {
  const nodeIds = await AsyncStorage.getItem(tile)
  return JSON.parse(nodeIds)
}

/**
 * Get node geometries for an array of tile ids
 * @param {Array<Object>} tiles
 * @return {<Object>}
 */
export async function getNodesForTiles (tiles) {
  const nodeIds = await AsyncStorage.multiGet(tiles)
  return nodeIds.reduce(async (memo, val) => {
    const newMemo = await Promise.resolve(memo)
    newMemo[val[0]] = await getNodes(JSON.parse(val[1]))
    return Promise.resolve({ ...newMemo })
  }, Promise.resolve({}))
}
