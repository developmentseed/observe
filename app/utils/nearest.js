/**
 * Find all nearest features from a given node
 */
import turfPointToLineDistance from '@turf/point-to-line-distance'
import turfDistance from '@turf/distance'
import { lineString } from '@turf/helpers'

export function findNearest (node, features) {
  const threshold = 0.02
  const nearestEdges = []
  const nearestNodes = []
  features.features.forEach(feature => {
    if (feature.geometry.type === 'LineString') {
      const distance = turfPointToLineDistance(node, feature, { 'units': 'kilometers' })
      if (distance < threshold) {
        feature.properties['distance'] = distance
        nearestEdges.push(feature)
      }
    }

    if (feature.geometry.type === 'Point') {
      const distance = turfDistance(node, feature, { 'units': 'kilometers' })
      if (distance < threshold) {
        nearestNodes.push(feature)
      }
    }

    if (feature.geometry.type === 'Polygon') {
      // get all the edges
      const edges = getEdge(feature)
      edges.forEach(edge => {
        const distance = turfPointToLineDistance(node, edge, { 'units': 'kilometers' })
        if (distance < threshold) {
          nearestEdges.push(edge)
        }
      })
    }
  })
  return {
    nearestEdges,
    nearestNodes
  }
}

function getEdge (polygon) {
  const edges = []
  const length = polygon.geometry.coordinates[0].length
  for (let index = 0; index < length - 1; index++) {
    const edge = polygon.geometry.coordinates[0].slice(index, index + 2)
    const edgeLine = lineString(edge, {})
    edges.push(edgeLine)
  }
  return edges
}
