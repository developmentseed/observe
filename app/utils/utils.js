export function isNewId (id) {
  id = String(id)
  return id.startsWith('observe-')
}
