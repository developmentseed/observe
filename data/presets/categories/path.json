{
    "icon": "temaki-pedestrian",
    "name": "Paths",
    "members": [
        "highway/path",
        "highway/footway",
        "highway/footway/marked",
        "highway/footway/sidewalk",
        "highway/steps",
        "highway/cycleway",
        "highway/bridleway",
        "highway/pedestrian_line"
    ]
}
