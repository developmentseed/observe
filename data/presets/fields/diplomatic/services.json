{
    "key": "diplomatic:services:",
    "type": "multiCombo",
    "label": "Services"
}
