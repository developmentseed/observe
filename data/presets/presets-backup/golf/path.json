{
    "icon": "maki-golf",
    "fields": [
        "name",
        "surface",
        "width",
        "structure",
        "incline"
    ],
    "geometry": [
        "line"
    ],
    "tags": {
        "golf": "path",
        "foot": "designated"
    },
    "reference": {
        "key": "golf",
        "value": "path"
    },
    "name": "Golf Walking Path"
}
