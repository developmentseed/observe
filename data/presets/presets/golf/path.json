{
    "icon": "maki-golf",
    "fields": [
        "name",
        "surface",
        "width",
        "structure",
        "incline"
    ],
    "geometry": [
        "line"
    ],
    "tags": {
        "golf": "path"
    },
    "addTags": {
        "golf": "path",
        "highway": "path",
        "foot": "designated"
    },
    "name": "Golf Walking Path"
}
