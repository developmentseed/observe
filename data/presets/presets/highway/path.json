{
    "icon": "iD-other-line",
    "fields": [
        "name",
        "surface",
        "width",
        "structure",
        "access",
        "incline"
    ],
    "moreFields": [
        "wheelchair",
        "lit",
        "smoothness",
        "trail_visibility",
        "sac_scale",
        "mtb/scale",
        "mtb/scale/uphill",
        "mtb/scale/imba",
        "horse_scale",
        "covered",
        "ref",
        "dog"
    ],
    "geometry": [
        "line"
    ],
    "terms": [
        "hike",
        "hiking",
        "trackway",
        "trail",
        "walk"
    ],
    "tags": {
        "highway": "path"
    },
    "name": "Path"
}
