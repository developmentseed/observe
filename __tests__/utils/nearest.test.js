/* global test */
/* global expect */
/* global __dirname */

import fs from 'fs'
import path from 'path'
import { findNearestPoint, findNearest } from '../../app/utils/nearest'
import { addNodes } from '../../app/services/nodecache'

test('find nearest features for a node', async () => {
  const features = JSON.parse(fs.readFileSync(path.join(__dirname, '../fixtures/osm-geojson-dc.geojson'), { 'encoding': 'utf-8' }))

  const nodes = JSON.parse(fs.readFileSync(path.join(__dirname, '../fixtures/dc-nodes.json'), { 'encoding': 'utf-8' }))
  const tile = '0320100322313212'
  await addNodes(tile, nodes)

  // node close to a polygon
  const node1 = {
    'type': 'Feature',
    'geometry': {
      'type': 'Point',
      'coordinates': [-77.0244, 38.9023]
    },
    'properties': {}
  }

  // node away from a linestring
  const node2 = {
    'type': 'Feature',
    'geometry': {
      'type': 'Point',
      'coordinates': [-77.0245, 38.9025]
    },
    'properties': {}
  }

  const nearest1 = await findNearest(node1, features)
  const nearest2 = await findNearest(node2, features)

  expect(nearest1).toMatchSnapshot()
  expect(nearest2).toMatchSnapshot()

  const edge = nearest1.nearestEdge
  const snappedPoint = findNearestPoint(node1, edge)
  expect(snappedPoint).toMatchSnapshot()
})
