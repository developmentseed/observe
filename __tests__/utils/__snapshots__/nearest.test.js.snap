// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`find nearest features for a node 1`] = `
Object {
  "nearestEdges": Array [
    Object {
      "geometry": Object {
        "coordinates": Array [
          Array [
            -77.0243478,
            38.902212,
          ],
          Array [
            -77.0243788,
            38.902298,
          ],
        ],
        "type": "LineString",
      },
      "properties": Object {
        "distance": 0.0018479514421559796,
      },
      "type": "Feature",
    },
    Object {
      "geometry": Object {
        "coordinates": Array [
          Array [
            -77.0243788,
            38.902298,
          ],
          Array [
            -77.025819,
            38.9022963,
          ],
        ],
        "type": "LineString",
      },
      "properties": Object {
        "distance": 0.000225172520188007,
      },
      "type": "Feature",
    },
  ],
  "nearestNodes": Array [
    Object {
      "geometry": Object {
        "coordinates": Array [
          -77.0243478,
          38.902212,
        ],
        "type": "Point",
      },
      "properties": Object {},
      "type": "Feature",
    },
    Object {
      "geometry": Object {
        "coordinates": Array [
          -77.0243788,
          38.902298,
        ],
        "type": "Point",
      },
      "properties": Object {},
      "type": "Feature",
    },
    Object {
      "geometry": Object {
        "coordinates": Array [
          -77.0243478,
          38.902212,
        ],
        "type": "Point",
      },
      "properties": Object {},
      "type": "Feature",
    },
    Object {
      "geometry": Object {
        "coordinates": Array [
          -77.0243788,
          38.902298,
        ],
        "type": "Point",
      },
      "properties": Object {},
      "type": "Feature",
    },
  ],
}
`;

exports[`find nearest features for a node 2`] = `
Object {
  "nearestEdges": Array [
    Object {
      "geometry": Object {
        "coordinates": Array [
          Array [
            -77.028073,
            38.9025312,
          ],
          Array [
            -77.0270204,
            38.9025427,
          ],
          Array [
            -77.026122,
            38.9025391,
          ],
          Array [
            -77.0239954,
            38.9025095,
          ],
        ],
        "type": "LineString",
      },
      "id": "way/4294967325",
      "properties": Object {
        "changeset": "17292",
        "distance": 0.0018370829928528922,
        "highway": "residential",
        "id": "way/4294967325",
        "ndrefs": Array [
          "4294967647",
          "4294969574",
          "4294967648",
          "4294967649",
        ],
        "timestamp": "2013-01-18T14:28:15Z",
        "uid": "15",
        "user": "TomH",
        "version": "2",
      },
      "type": "Feature",
    },
  ],
  "nearestNodes": Array [],
}
`;
